### PR TITLE
[TSv3] Step 0 - Compat layer for using v3 training service as v1

### PR DESCRIPTION
**Pull request by liuzhe-lz**
### Description ###

This is the preparation of training service v3.

It contains the new interface and one-side compatibility layer.

The compat layer allows to use a v3 training service as a v1 training service, so we can test new training services without modifying NNI manager.

#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [ ] full test - compression

### Checklist ###
  - [ ] test case
  - [ ] doc

### CODE REVIEW

File: ts/nni_manager/common/training_service_v3.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    init(): Promise<void>;
+    start(): Promise<EnvironmentInfo[]>;
+    stop(): Promise<void>;
+
+    uploadDirectory(directoryName: string, path: string): Promise<void>;
+    createTrial(environmentId: string, trialCommand: string, directoryName: string): Promise<string | null>;
+    stopTrial(trialId: string): Promise<void>;
+    sendParameter(trialId: string, parameter: Parameter): Promise<void>;
+
+    onRequestParameter(callback: (trialId: string) => Promise<void>): void;
```
Comment (ultmaster) [inline ts/nni_manager/common/training_service_v3.ts:32]: One trial can still have multiple parameter?

File: ts/nni_manager/common/training_service_v3.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    id: string;
+}
+
+export interface TrainingServiceV3 {
+    init(): Promise<void>;
+    start(): Promise<EnvironmentInfo[]>;
+    stop(): Promise<void>;
+
+    uploadDirectory(directoryName: string, path: string): Promise<void>;
+    createTrial(environmentId: string, trialCommand: string, directoryName: string): Promise<string | null>;
```
Comment (ultmaster) [inline ts/nni_manager/common/training_service_v3.ts:28]: Would `createTrial` send a parameter with it?

File: ts/nni_manager/common/training_service_v3.ts (hunk 1 above)
Comment (liuzhe-lz) [reply ts/nni_manager/common/training_service_v3.ts:74]: Yes.
Each `get_next_parameter()` results in one `onRequestParameter()`, and then one `sendParameter()`.

File: ts/nni_manager/common/training_service_v3.ts (hunk 2 above)
Comment (liuzhe-lz) [reply ts/nni_manager/common/training_service_v3.ts:62]: By design, no.
If the trial does not invoke `get_next_parameter()` or `get_current_parameter()`, no parameter will be generated.

In current implementation, yes.
The compat layer caches a parameter on creating trial, and send it to first `onRequestParameter()`.

File: ts/nni_manager/common/training_service_v3.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+export interface TrainingServiceV3 {
+    /**
+     *  Invoked during experiment initialization.
+     *
+     *  It should verify the config and raise error if the specified training resource is not available.
+     *
+     *  It should not start daemon on worker machine.
+     *  If another component of the experiment failed to initialize,
+     *  the process will abort without invoking any clean up function.
```
Comment (QuanluZhang) [inline ts/nni_manager/common/training_service_v3.ts:31]: this implies that there should be clean-up but you decide not to do the clean-up, why?

**Review by QuanluZhang — APPROVED** [review]

**Review by ultmaster — APPROVED** [review]

File: ts/nni_manager/common/training_service_v3.ts (hunk 3 above)
Comment (liuzhe-lz) [reply ts/nni_manager/common/training_service_v3.ts:31]: This method is invoked during initializing stage.
Failing to initialize means NNI manager is in an unexpected condition, so nothing is guaranteed to work, and halting is the safest action.

**Pull request merged**